src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -873,19 +873,26 @@ DagState uniqifyState(ImmutableArray<StateForCase> cases, ImmutableDictionary<Bo
                             break;
                         case BoundDagTest d:
                             bool foundExplicitNullTest = false;
+                            bool foundExplicitValueTest = false;
                             SplitCases(state, d,
                                 out ImmutableArray<StateForCase> whenTrueDecisions,
                                 out ImmutableArray<StateForCase> whenFalseDecisions,
                                 out ImmutableDictionary<BoundDagTemp, IValueSet> whenTrueValues,
                                 out ImmutableDictionary<BoundDagTemp, IValueSet> whenFalseValues,
-                                ref foundExplicitNullTest);
+                                ref foundExplicitNullTest, ref foundExplicitValueTest);
                             state.TrueBranch = uniqifyState(whenTrueDecisions, whenTrueValues);
                             state.FalseBranch = uniqifyState(whenFalseDecisions, whenFalseValues);
                             if (foundExplicitNullTest && d is BoundDagNonNullTest { IsExplicitTest: false } t)
                             {
                                 // Turn an "implicit" non-null test into an explicit one
                                 state.SelectedTest = new BoundDagNonNullTest(t.Syntax, isExplicitTest: true, t.Input, t.HasErrors);
                             }
+                            else if (foundExplicitValueTest && d.WasCompilerGenerated && d is BoundDagValueTest v)
+                            {
+                                // We're going to remove compiler-generated nodes from dag right after construction.
+                                // If we have found an explicit value test we need to unset the flag to preserve it.
+                                state.SelectedTest = new BoundDagValueTest(v.Syntax, v.Value, v.Input, v.HasErrors);
```

Comment:
>state.SelectedTest = new BoundDagValueTest(v.Syntax, v.Value, v.Input, v.HasErrors); [](http://example.com/codeflow?start=32&length=84)

I am assuming we have tests that confirm significance of this operation. Could you provide an example of an affected scenario and how the decompiled code looks like for it?

Reply:
The compiler-generated length test (L=1) is merged with the length test from the second pattern.
 ```cs
           switch (a)
            {
                case [.., 1]:
                case [2]:
                    return 0;
            }
```
```cs
            if (a != null)
            {
                int length = a.Length;
                if (length >= 1 && (a[length - 1] == 1 || (length == 1 && a[0] == 2)))
                {
                    return 0;
                }
            }
```

Reply:
> Not doing so would eliminate the whenTrue branch and cause a false subsumption error.

I thought ```WasCompilerGenerated``` flag didn't have any impact on subsumption checking. Is this incorrect?

Reply:
Can you think of a scenario when leaving this node won't be useful at the end? For example, when ```length``` is explicitly checked. Something like:
```
switch (a)
{
    case [.., 1]:
    case [2, ..] and (Length: <3 or > 5):
        return 0;
}
```

Reply:
It doesn't, but since we're removing nodes based on it, it should not be set where it's not supposed to.
I'm gonna go ahead and make the change to not depend on WasCompilerGenerated. I think that'll make all this more clear.

Reply:
>a scenario when leaving this node won't be useful at the end? 

In your example the node is flagged and will be removed (there's no matching length value test)

As long as this analysis goes, it doesn't matter where the the second length check comes from. We're only interested in an exact value test because we're only inserting an exact value test.

Reply:
The node is only preserved when there's a matching value test which is not subsumption-only.

Reply:
> In your example the node is flagged and will be removed (there's no matching length value test)

Ok, that is a bad example. However, this doesn't answer my question: "Can you think of a scenario when leaving this node won't be useful at the end?" Actually, I'll try to rephrase the question to better reflect my concern. Can you think of a scenario where, leaving any artifacts of subsumption checking in the DAG (including addition of checks and merging happening because of that) leads to a suboptimal code by comparison to the code we would produce if we didn't do any list pattern specific subsumption checking? In other words. Imagine we built one DAG as we do in this PR and we built another one with the list pattern subsumption checks disabled (we do not synthesize any nodes that are subject to elimination). Can you think of a scenario for which the second DAG is "better" or more optimal than the first one?

Reply:
>a scenario when leaving this node won't be useful at the end?

I think you mean *if* there is a matching length value test (from a list or otherwise) is there a scenario that could leave the node in the final dag where it's not actually useful.

I couldn't think of such scenario. if the value is the same it gets merged, otherwise the dag is split where we left with unrelated branches in which case we won't reset the flag and it gets removed. Does that make sense?

Reply:
Ok that was my speculation. Let me see if I can come up with a concrete example.

Reply:
A detail I admittedly forgot about is that we only insert a length test for two indexers that could relate. So for a length test alone we don't have to deal with a synthesized test. That got me confused.

>Imagine we built one DAG as we do in this PR and we built another one with the list pattern subsumption checks disabled (we do not synthesize any nodes that are subject to elimination). Can you think of a scenario for which the second DAG is "better" or more optimal than the first one?

This change is meant to modify the DAG such that we never inserted such nodes.

Notice how we rewrite tests in each case: `((P && S) || other)`, `(!(P && S) && other)`, `((!P || S) && other)`
When there is no precondition (P=True) those will fallback to the original behavior: `True`, `False`, and `other`.
When there is a precondition, originally tests were deemed to be unrelated (no rewriting)
Substitute (P=False) and you will get `other` which is the same node we'd have with list pattern subsumption checks disabled.

Reply:
Sorry, this thread has become too long already, but there is still something I need to get clarity on. Obviously, there is a clear benefit in having subsumption checking and I am not suggesting to remove it. But I still don't have a clear understanding whether leaving any artifacts of subsumption checking in the Dag benefits the code gen at the end. Do we have any test case that clearly demonstrates the benefit? If those artifacts don't provide any benefit, then I would feel more comfortable with a solution that doesn't preserve any subsumption checking artifacts in the final Dag, if we could find it.

Reply:
>I would feel more comfortable with a solution that doesn't preserve any subsumption checking artifacts in the final Dag

That's the goal. A subsumption-only test should be preserved only if it's going to be merged with an existing node in which case it's not really an artifact (shown in the example at the top of the thread). 

That doesn't actually *add* anything to the DAG - the only thing we have on the true branch is an skipped evaluation, so no impact on the codegen.

We should just make sure we don't unset the flag where it's not absolutely necessary (see comment below).

Reply:
The reason why we need to do this in the first place, is that a synthesized length test may come *before* any existing one (if any). In case we unset the flag, it will cause a reordering (evaluating the test earlier than we normally do). I don't thing that's avoidable at moment, but I'll need to investigate to make sure the impact is at least acceptable.

Reply:
I think I am failing to get my point across. By my definition, any modification to the Dag made due to injection of synthesized nodes created for the purpose of subsumption checking is an artifact that concerns me. I understand that an injection of synthesized nodes might lead to other mutations in the Dag (merging, etc.) I also understand, that here we are trying ensure correctness by preserving a synthesized node that caused mutation of the Dag. This is still an artifact by my definition. If such an artifact doesn't provide any advantage to the quality of the generated code, I think it is better to generate code based on the Dag that doesn't have any artifacts in it. As if no synthesized nodes were injected in the first place.

Reply:
To be clear, are you suggesting we should also avoid the reordering of the length test? (assuming that would be the only artifact after this change)

Reply:
> To be clear, are you suggesting we should also avoid the possible reordering of the length test? 

Yes, unless it provides clear benefits to the quality of the code we generate.

> assuming that would be the only artifact after this change

This isn't quite obvious to me. Aren't we are going to keep the synthetic branch that comes along with the synthetic length test? The branch effectively duplicates some portion of logic.

Reply:
@AlekseyTs 

>The branch effectively duplicates some portion of logic.

<details><summary> Here's the same example in commit 8 </summary>

```cs
    private static int Test1(int[] a)
    {
        int num;
        if (a != null)
        {
            int length = a.Length;
            if (length >= 2)
            {
                if (a[length - 2] >= 0)
                {
                    if (a[length - 1] > 0)
                    {
                        return 1;
                    }
                    num = a[0];
                    if (length != 2)
                    {
                        goto IL_004e;
                    }
                    if (num > 0)
                    {
                        int num2 = a[1];
                        goto IL_0068;
                    }
                    if (a[1] >= 0)
                    {
                        return 5;
                    }
                }
                else
                {
                    if (a[length - 1] >= 0)
                    {
                        return 2;
                    }
                    num = a[0];
                    if (length != 2)
                    {
                        goto IL_004e;
                    }
                    int num3 = a[1];
                }
                goto IL_0066;
            }
            goto IL_006c;
        }
        return 7;
        IL_0068:
        return 4;
        IL_0066:
        return 3;
        IL_006c:
        return 6;
        IL_004e:
        if (num <= 0)
        {
            if (a[1] < 0)
            {
                goto IL_0066;
            }
        }
        else if (a[1] <= 0)
        {
            goto IL_0068;
        }
        goto IL_006c;
    }
```

</details>

I don't think we can avoid duplicated nodes. Ignoring unused evaluations, what other improvements you think could be made to this code? If you believe we should emit code as good as the one with subsumption-checking disabled, we might as well construct a new dag dedicated to lowering.

Reply:
>I don't think we can avoid duplicated nodes. Ignoring unused evaluations, what other improvements you think could be made to this code?

It is quite possible that the scenario doesn't lead to the duplication because one of the branches is optimized away. Basically we determined that either TrueBranch or FalseBranch is definitely false. However, I don't see anything in the implementation that would guarantee that we never end up with two branches alive.

To clarify, I am not looking for ways to improve the code. I am looking for a definitive proof that the code is "better" than the one we would produce without the list pattern subsumption checks in place. This means that we have to compare two code gen strategies and see that the one with artifacts is definitely better. Perhaps even to the point that we would want to implement it if the subsumption checking wouldn't give it to us for free. If there is no proof like that, then I think that it is better to generate code from a Dug that doesn't have any artifact. Instead of marking things and then hoping they can be removed safely and hoping for the best if they couldn't be removed.

>we might as well construct a new dag dedicated to lowering.

This is definitely an option if we cannot find a way to achieve the same in some "incremental" fashion.

Reply:
> I don't see anything in the implementation that would guarantee that we never end up with two branches alive.

I think the fact that we can't optimally handle such branches is a general issue with the existing pattern-matching machinery. We use a simple left-to-right heuristic which may not result in optimal code with certain trees. See https://github.com/dotnet/roslyn/issues/29033.

Reply:
> One possible definition of "better code" would be "fewer number of tests before we jump to a leaf node" and by this definition we're emitting a better code. 

This is a theory or a hope. This must be proven.

> I think the fact that we can't optimally handle such branches is a general issue with the existing pattern-matching machinery. We use a simple left-to-right heuristic which may not result in optimal code with certain trees. See [https://github.com/dotnet/roslyn/issues/29033](https://github.com/dotnet/roslyn/issues/29033).

Again, I am not going for the most optimal code gen. I simply would like to see a confirmation that there is a good reason to keep any artifacts of the list pattern subsumption checks in the generated code. We are complicating implementation, making an assumption that we are able to accurately detect "unremovable" artifacts. We are also complicating the future maintenance of the machinery, we have to make sure that future changes do not invalidate the assumption, etc. All this comes with additional risk and cost. Why do we want to take it? Is there a good reason for that?

Reply:
Simply removing *all* the artifacts won't work as they're practically part of the input now and the resulting dag depends on them to occur at some point during execution. We can, however, short-circuit some of them that aren't crucial to the result.


I think we have two options moving forward:

1. Move the logic to detect unavoidable artifacts to a later step maybe based on dag nodes themselves instead of looking for a "matching source test" which I take it is vaguely defined.
2. Re-compute the dag for lowering. Maybe we can reuse some of the nodes but to my understanding, we can't do it as a "modification" to the existing tree. The two could be very different in shape to be able to "translate back" to the original.

Reply:
> What is your recommendation on approaching this?

Thank you, @alrz. We will discuss this internally and will get back to you early next week.